From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HAOS Fork <haos@local>
Date: Sat, 15 Mar 2026 00:00:00 +0000
Subject: [PATCH] arm64: dts: amlogic: add X88 Pro X3 device tree

Add device tree support for the X88 Pro X3 (Amlogic S905X3) TV box.
Based on the unifreq community kernel tree. Uses the mainline
meson-sm1-ac2xx.dtsi as base.

Signed-off-by: HAOS Fork <haos@local>
---
 arch/arm64/boot/dts/amlogic/Makefile                     | 2 ++
 arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3.dts    | 90 ++++++++++
 arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3-4g.dts | 18 ++++++++++
 3 files changed, 110 insertions(+)
 create mode 100644 arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3.dts
 create mode 100644 arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3-4g.dts

diff --git a/arch/arm64/boot/dts/amlogic/Makefile b/arch/arm64/boot/dts/amlogic/Makefile
--- a/arch/arm64/boot/dts/amlogic/Makefile
+++ b/arch/arm64/boot/dts/amlogic/Makefile
@@ -92,4 +92,6 @@
 dtb-$(CONFIG_ARCH_MESON) += meson-sm1-x96-air-gbit.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-sm1-x96-air.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson-sm1-x88-pro-x3.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson-sm1-x88-pro-x3-4g.dtb
 
 meson-sm1-khadas-vim3l-ts050-dtbs:= meson-sm1-khadas-vim3l.dtb meson-khadas-vim3-ts050.dtbo
diff --git a/arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3.dts b/arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3.dts
new file mode 100644
--- /dev/null
+++ b/arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3.dts
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2023 AI-Speaker.com. All rights reserved.
+ */
+
+/dts-v1/;
+
+#include "meson-sm1-ac2xx.dtsi"
+#include <dt-bindings/sound/meson-g12a-tohdmitx.h>
+
+/ {
+	compatible = "ai-speaker.com,dev3\0amlogic,sm1";
+	model = "AIS DEV3 / X88 Pro Pro X3 / X99 Max Plus / Transpeed X3 Plus";
+
+	fd655_dev {
+		dat_pin = <0x18 0x40 0x00>;
+		dev_name = "fd655_dev";
+		status = "okay";
+		clk_pin = <0x18 0x41 0x00>;
+		compatible = "amlogic,fd655_dev";
+	};
+
+	openvfd {
+		compatible = "open,vfd";
+		dev_name = "openvfd";
+		status = "okay";
+	};
+};
+
+
+&ethmac {
+	status = "okay";
+	pinctrl-0 = <&eth_pins>, <&eth_rgmii_pins>;
+	pinctrl-names = "default";
+	phy-handle = <&external_phy>;
+	phy-mode = "rgmii";
+	amlogic,tx-delay-ns = <2>;
+	snps,aal;
+	snps,rxpbl = <0x8>;
+	snps,txpbl = <0x8>;
+	rx-fifo-depth = <4096>;
+	tx-fifo-depth = <4096>;
+	nvmem-cells = <&eth_mac>;
+	nvmem-cell-names = "eth_mac";
+};
+
+&ext_mdio {
+	external_phy: ethernet-phy@0 {
+		/* Realtek RTL8211F */
+		reg = <0>;
+		max-speed = <1000>;
+		reset-assert-us = <10000>;
+		reset-deassert-us = <80000>;
+		reset-gpios = <&gpio GPIOZ_15 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>;
+
+		interrupt-parent = <&gpio_intc>;
+		/* MAC_INTR on GPIOZ_14 */
+		interrupts = <IRQID_GPIOZ_14 IRQ_TYPE_LEVEL_LOW>;
+	};
+};
+
+&efuse {
+	eth_mac: eth_mac@0 {
+		reg = <0x0 0x06>;
+	};
+
+	bt_mac: bt_mac@6 {
+		reg = <0x6 0x06>;
+	};
+
+	wifi_mac: wifi_mac@12 {
+		reg = <0x0c 0x06>;
+	};
+
+	usid: usid@18 {
+		reg = <0x12 0x10>;
+	};
+};
+
+// set max cpu speed
+&cpu_opp_table {
+
+	opp-2016000000 {
+		status = "disabled";
+	};
+
+	opp-2100000000 {
+		status = "disabled";
+	};
+};
diff --git a/arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3-4g.dts b/arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3-4g.dts
new file mode 100644
--- /dev/null
+++ b/arch/arm64/boot/dts/amlogic/meson-sm1-x88-pro-x3-4g.dts
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2023 AI-Speaker.com. All rights reserved.
+ */
+
+/dts-v1/;
+
+#include "meson-sm1-x88-pro-x3.dts"
+
+/ {
+	compatible = "amlogic,sm1";
+	model = "X88-Pro-X3 4G";
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x0 0x1 0x0>; // 4GB
+	};
+};
-- 
